ocaml/xapi/storage_impl.ml

```diff
@@ -568,26 +573,29 @@ functor
           )
         in
         let failures =
-          List.fold_left
-            (fun acc dp ->
-              info "Attempting to destroy datapath dp:%s sr:%s vdi:%s" dp
-                (s_of_sr sr) (s_of_vdi vdi) ;
-              try
-                destroy_datapath_nolock context ~dbg ~dp ~sr ~vdi ~vm
-                  ~allow_leak:false ;
-                acc
-              with e -> e :: acc
-            )
-            [] dps
+          if fail_on_leaked && dps <> [] then
+            ["Leaked Datapaths not allowed"]
+          else
+            List.fold_left
+              (fun acc (dpid, vm) ->
+                info "Attempting to destroy datapath dp:%s sr:%s vdi:%s" dpid
+                  (s_of_sr sr) (s_of_vdi vdi) ;
+                try
+                  destroy_datapath_nolock context ~dbg ~dp:dpid ~sr ~vdi ~vm
+                    ~allow_leak:false ;
+                  acc
+                with e -> Printexc.to_string e :: acc
```

Comment:
Now you are rewriting exceptions and always raise `Failure` rather than the original exception, which could've been a different one...